Fix broadcasting of cdist backward

albanD opened the pull request:
Fix https://github.com/pytorch/pytorch/issues/55370


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56606 Fix cdist backward for empty inputs
* **#56605 Fix broadcasting of cdist backward**
* #56604 Add more test cases for cdist OpInfo and TODOs
* #56603 Add option to OpInfo to skip gradgrad check and empty cdist OpInfo

Differential Revision: [D27939202](https://our.internmc.facebook.com/intern/diff/D27939202)

albanD committed:
Fix broadcasting of cdist backward

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -145,7 +145,31 @@ Tensor _cdist_forward(const Tensor& x1, const Tensor& x2, const double p, c10::o
+  auto dim2 = _x2.dim();
+  IntArrayRef batch_tensor1(_x1.sizes().data(), dim1 - 2);
+  IntArrayRef batch_tensor2(_x2.sizes().data(), dim2 - 2);
+  std::vector<int64_t> expand_batch_portion = infer_size(batch_tensor1, batch_tensor2);
```
I could tell you to use infer_size_dimvector here to avoid std::vector, and use DimVectors for you expand sizes below, and use MaybeOwned to avoid refcount bumps with your maybe expands, but cdist backward is so awful that none of this would matter.

albanD replied:
Ok leaving it like that for now then.

ngimel commented:
hm, test_variant_consistency still fails with wrong size gradients.

albanD commented:
Yes I didn't run the cuda tests locally, so I was expecting something like that...
Time to do a local cuda build I guess :D

albanD committed:
Update on "Fix broadcasting of cdist backward"

Fix https://github.com/pytorch/pytorch/issues/55370





[ghstack-poisoned]

albanD commented:
@ngimel updated this PR and the one above to fix also the CUDA implementation that was suffering from similar issues.

ngimel commented:
Looks good, let's see what tests say.

albanD committed:
Update on "Fix broadcasting of cdist backward"

Fix https://github.com/pytorch/pytorch/issues/55370



Differential Revision: [D27939202](https://our.internmc.facebook.com/intern/diff/D27939202)

[ghstack-poisoned]

albanD committed:
Update on "Fix broadcasting of cdist backward"

Fix https://github.com/pytorch/pytorch/issues/55370



Differential Revision: [D27939202](https://our.internmc.facebook.com/intern/diff/D27939202)

[ghstack-poisoned]
